File: clients/android/core/src/main/kotlin/com/heibaimiao/multilivetv/source/SourceAdapter.kt
package com.heibaimiao.multilivetv.source

import com.heibaimiao.multilivetv.maccms.MacCMSClient
import com.heibaimiao.multilivetv.maccms.VodTypeRaw
import com.heibaimiao.multilivetv.model.Source
import com.heibaimiao.multilivetv.net.VodClientException

interface SourceAdapter {
    suspend fun list(source: Source, page: Int, typeId: Int?, hours: Int?): SourcePage
    suspend fun search(source: Source, keyword: String, page: Int): SourcePage
    suspend fun detail(source: Source, sourceMovieId: String): SourceMovie
    suspend fun fetchTypes(source: Source): List<VodTypeRaw>
}

object SourceAdapterRegistry {
    fun adapter(source: Source): SourceAdapter = when (source.adapter.type) {
        "cms_json" -> CMSJsonAdapter
        else -> throw VodClientException("不支持的 Adapter: ${source.adapter.type}")
    }
}

object SourceCollector {
    suspend fun list(
        source: Source,
        page: Int,
        typeId: Int?,
        hours: Int? = null,
        health: SourceHealthStore = SourceHealthStore.shared,
    ): SourcePage {
        if (!source.enabled) throw VodClientException("源不支持能力: enabled")
        if (!source.capabilities.category) throw VodClientException("源不支持能力: category")
        val adapter = SourceAdapterRegistry.adapter(source)
        return try {
            adapter.list(source, page, typeId, hours).also {
                health.recordSuccess(source.numericId)
            }
        } catch (error: Exception) {
            health.recordFailure(source.numericId)
            throw error
        }
    }

    suspend fun search(
        source: Source,
        keyword: String,
        page: Int,
        health: SourceHealthStore = SourceHealthStore.shared,
    ): SourcePage {
        if (!source.enabled) throw VodClientException("源不支持能力: enabled")
        if (!source.capabilities.search) throw VodClientException("源不支持能力: search")
        val adapter = SourceAdapterRegistry.adapter(source)
        return try {
            adapter.search(source, keyword, page).also {
                health.recordSuccess(source.numericId)
            }
        } catch (error: Exception) {
            health.recordFailure(source.numericId)
            throw error
        }
    }

    suspend fun detail(
        source: Source,
        sourceMovieId: String,
        health: SourceHealthStore = SourceHealthStore.shared,
    ): SourceMovie {
        if (!source.enabled) throw VodClientException("源不支持能力: enabled")
        if (!source.capabilities.detail) throw VodClientException("源不支持能力: detail")
        val adapter = SourceAdapterRegistry.adapter(source)
        return try {
            adapter.detail(source, sourceMovieId).also {
                health.recordSuccess(source.numericId)
            }
        } catch (error: Exception) {
            health.recordFailure(source.numericId)
            throw error
        }
    }

    suspend fun fetchTypes(source: Source): List<VodTypeRaw> {
        if (!source.enabled) throw VodClientException("源不支持能力: enabled")
        return SourceAdapterRegistry.adapter(source).fetchTypes(source)
    }
}

object CMSJsonAdapter : SourceAdapter {
    override suspend fun list(source: Source, page: Int, typeId: Int?, hours: Int?): SourcePage {
        val effectivePage = if (source.capabilities.pagination) page else 1
        val response = MacCMSClient.fetchList(source, effectivePage, typeId, hours)
        return SourcePage(
            page = response.page,
            pageCount = response.pageCount,
            total = response.total,
            list = response.list.map { MacCMSSourceParser.toSourceMovie(source, it) },
        )
    }

    override suspend fun search(source: Source, keyword: String, page: Int): SourcePage {
        val effectivePage = if (source.capabilities.pagination) page else 1
        val response = MacCMSClient.search(source, keyword, effectivePage)
        return SourcePage(
            page = response.page,
            pageCount = response.pageCount,
            total = response.total,
            list = response.list.map { MacCMSSourceParser.toSourceMovie(source, it) },
        )
    }

    override suspend fun detail(source: Source, sourceMovieId: String): SourceMovie {
        val response = MacCMSClient.fetchDetail(source, sourceMovieId)
        val raw = response.list.firstOrNull() ?: throw VodClientException("详情为空")
        return MacCMSSourceParser.toSourceMovie(source, raw)
    }

    override suspend fun fetchTypes(source: Source): List<VodTypeRaw> = MacCMSClient.fetchTypes(source)
}

File: clients/apple/MultiLiveTV/Services/SourceAdapter.swift
import Foundation

protocol SourceAdapter {
    func list(source: Source, page: Int, typeId: Int?, hours: Int?) async throws -> SourcePage
    func search(source: Source, keyword: String, page: Int) async throws -> SourcePage
    func detail(source: Source, sourceMovieId: String) async throws -> SourceMovie
    func fetchTypes(source: Source) async throws -> [VodTypeRaw]
}

enum SourceAdapterError: LocalizedError {
    case unsupportedAdapter(String)
    case capabilityDisabled(String)
    case emptyDetail

    var errorDescription: String? {
        switch self {
        case .unsupportedAdapter(let type):
            return "不支持的 Adapter: \(type)"
        case .capabilityDisabled(let name):
            return "源不支持能力: \(name)"
        case .emptyDetail:
            return "详情为空"
        }
    }
}

enum SourceAdapterRegistry {
    static func adapter(for source: Source) throws -> SourceAdapter {
        switch source.adapter.type {
        case "cms_json":
            return CMSJsonAdapter.shared
        default:
            throw SourceAdapterError.unsupportedAdapter(source.adapter.type)
        }
    }
}

/// Capability-gated collector used by list / search / detail fan-out.
enum SourceCollector {
    static func list(
        source: Source,
        page: Int,
        typeId: Int?,
        hours: Int? = nil,
        health: SourceHealthStore = .shared
    ) async throws -> SourcePage {
        guard source.enabled else { throw SourceAdapterError.capabilityDisabled("enabled") }
        guard source.capabilities.category else { throw SourceAdapterError.capabilityDisabled("category") }
        let adapter = try SourceAdapterRegistry.adapter(for: source)
        do {
            let pageResult = try await adapter.list(source: source, page: page, typeId: typeId, hours: hours)
            health.recordSuccess(numericId: source.numericId)
            return pageResult
        } catch {
            health.recordFailure(numericId: source.numericId)
            throw error
        }
    }

    static func search(
        source: Source,
        keyword: String,
        page: Int,
        health: SourceHealthStore = .shared
    ) async throws -> SourcePage {
        guard source.enabled else { throw SourceAdapterError.capabilityDisabled("enabled") }
        guard source.capabilities.search else { throw SourceAdapterError.capabilityDisabled("search") }
        let adapter = try SourceAdapterRegistry.adapter(for: source)
        do {
            let pageResult = try await adapter.search(source: source, keyword: keyword, page: page)
            health.recordSuccess(numericId: source.numericId)
            return pageResult
        } catch {
            health.recordFailure(numericId: source.numericId)
            throw error
        }
    }

    static func detail(
        source: Source,
        sourceMovieId: String,
        health: SourceHealthStore = .shared
    ) async throws -> SourceMovie {
        guard source.enabled else { throw SourceAdapterError.capabilityDisabled("enabled") }
        guard source.capabilities.detail else { throw SourceAdapterError.capabilityDisabled("detail") }
        let adapter = try SourceAdapterRegistry.adapter(for: source)
        do {
            let movie = try await adapter.detail(source: source, sourceMovieId: sourceMovieId)
            health.recordSuccess(numericId: source.numericId)
            return movie
        } catch {
            health.recordFailure(numericId: source.numericId)
            throw error
        }
    }

    static func fetchTypes(source: Source) async throws -> [VodTypeRaw] {
        guard source.enabled else { throw SourceAdapterError.capabilityDisabled("enabled") }
        let adapter = try SourceAdapterRegistry.adapter(for: source)
        return try await adapter.fetchTypes(source: source)
    }
}

struct CMSJsonAdapter: SourceAdapter {
    static let shared = CMSJsonAdapter()

    func list(source: Source, page: Int, typeId: Int?, hours: Int?) async throws -> SourcePage {
        let effectivePage = source.capabilities.pagination ? page : 1
        let response = try await MacCMSClient.fetchList(
            source: source,
            page: effectivePage,
            typeId: typeId,
            hours: hours
        )
        return SourcePage(
            page: response.page,
            pageCount: response.pageCount,
            total: response.total,
            list: response.list.map { MacCMSSourceParser.toSourceMovie(source: source, raw: $0) }
        )
    }

    func search(source: Source, keyword: String, page: Int) async throws -> SourcePage {
        let effectivePage = source.capabilities.pagination ? page : 1
        let response = try await MacCMSClient.search(source: source, keyword: keyword, page: effectivePage)
        return SourcePage(
            page: response.page,
            pageCount: response.pageCount,
            total: response.total,
            list: response.list.map { MacCMSSourceParser.toSourceMovie(source: source, raw: $0) }
        )
    }

    func detail(source: Source, sourceMovieId: String) async throws -> SourceMovie {
        let response = try await MacCMSClient.fetchDetail(source: source, ids: sourceMovieId)
        guard let raw = response.list.first else { throw SourceAdapterError.emptyDetail }
        return MacCMSSourceParser.toSourceMovie(source: source, raw: raw)
    }

    func fetchTypes(source: Source) async throws -> [VodTypeRaw] {
        try await MacCMSClient.fetchTypes(source: source)
    }
}
